### DISCUSSION TITLE
Bug report for U-boot

### DISCUSSION THREAD

**Discussion by ltrsunny** [linux-chenxing/linux-chenxing.org#49] [General]
Branch is mstar_rebase_mainline

```
In file included from cmd/gpio.c:15:
./arch/arm/include/asm/gpio.h:7:10: fatal error: asm/arch/gpio.h: No such file or directory
    7 | #include <asm/arch/gpio.h>
```

**Comment by fifteenhex** [comment]
Which config did you use? The u-boot port is a mess at the moment.

**Reply by ltrsunny** [reply]
> Which config did you use? The u-boot port is a mess at the moment.

mstar_infinity2m_som2d01_defconfig

I manually enabled gpio command, because I want to use io to select different boot modes.

**Reply by fifteenhex** [reply]
I don't think I've written a gpio driver for u-boot yet. I haven't needed it so far.

**Reply by ltrsunny** [reply]
Okay……

**Reply by fifteenhex** [reply]
Writing a GPIO driver for u-boot wouldn't be that difficult. Do you think you could give it a try? Otherwise just poke the gpio registers with mw.w xxx xxx until I get time to do it.

**Reply by ltrsunny** [reply]
I'd like to! But let me try to make a usable bootloader first……

**Comment by ltrsunny** [comment]
Seems USB and MMC not working as well.
```
arm-linux-gnu-ld.bfd: drivers/built-in.o: in function `mstar_mmc_bind':
/home/dev/mstar-uboot/drivers/mmc/mstar_mmc.c:555: undefined reference to `mmc_bind'
```
```
arm-linux-gnu-ld.bfd: drivers/usb/gadget/udc/built-in.o: in function `usb_gadget_handle_interrupts':
/home/dev/mstar-uboot/drivers/usb/gadget/udc/udc-uclass.c:56: undefined reference to `dm_usb_gadget_handle_interrupts'
```

**Reply by fifteenhex** [reply]
u-boot has a usb driver for the same controller but it needs to be wired up.
MMC should work.. I've been using it to boot other boards. It looks like you need to select another option to compile in the mmc support to u-boot

**Reply by ltrsunny** [reply]
I enabled  DM_MMC and MMC_MSTAR, There is a warning at the beginning of compilation:
```
===================== WARNING ======================
This board does not use CONFIG_DM_MMC. Please update
the board to use CONFIG_DM_MMC before the v2019.04 release.
Failure to update by the deadline may result in board removal.
See doc/driver-model/migration.rst for more info.
====================================================
```

do you have any idea?

**Reply by fifteenhex** [reply]
I think that might be safe to ignore. Our code isn't in mainline u-boot yet so it's not like they can remove us :D

**Reply by ltrsunny** [reply]
But the error is still there.



> ```
> arm-linux-gnu-ld.bfd: drivers/built-in.o: in function `mstar_mmc_bind':
> /home/dev/mstar-uboot/drivers/mmc/mstar_mmc.c:555: undefined reference to `mmc_bind'
> ```
> ```
> ```

**Reply by ltrsunny** [reply]
The function`mmc_bind()` is defined in `drivers/mmc/mmc-uclass.c`

**Reply by fifteenhex** [reply]
Looks like you need CONFIG_BLK

**Reply by ltrsunny** [reply]
> Looks like you need CONFIG_BLK

Yes！

**Reply by ltrsunny** [reply]
Looping when using mmcinfo
```
timeout while polling
timeout while polling
timeout while polling
timeout while polling
timeout while polling
timeout while polling
```
keep debugging…

**Reply by fifteenhex** [reply]
I think the ssd202d might need a fix for the extra clock mux/gate .. I thought that was in u-boot but maybe not.
I'll take a look at it tomorrow. I guess you want to put your kernel on the SD card?

**Reply by ltrsunny** [reply]
> I think the ssd202d might need a fix for the extra clock mux/gate .. I thought that was in u-boot but maybe not. I'll take a look at it tomorrow. I guess you want to put your kernel on the SD card?

Yes, I want to make kernel and rootfs upgradable via USB, that is why I am also trying GPIO and USB. Thanks a lot.

**Reply by ltrsunny** [reply]
@fifteenhex If you are quite busy now, just give me some tips and let me try.

**Reply by fifteenhex** [reply]
Give me a few hours. We need to get our changes onto the latest u-boot before doing any new work.
I'm working on that now. The next thing will be to fix the missing clock (I think) for SD to work.
I need to check what state the USB driver is in and see if we can fix that.

**Reply by ltrsunny** [reply]
> Give me a few hours. We need to get our changes onto the latest u-boot before doing any new work. I'm working on that now. The next thing will be to fix the missing clock (I think) for SD to work. I need to check what state the USB driver is in and see if we can fix that.

Thank you very much. I also want to contribute something if I can.

**Reply by fifteenhex** [reply]
> Thank you very much. I also want to contribute something if I can.

Once I have our u-boot updated to the current master you can try writing the GPIO driver. I don't think it'll be too difficult and I can help out.
There isn't much point starting that now as a lot of things have changed since 2021.01 (our version) so you'd end up having to do fixes to get it to work again after we are running master.

I'm close to getting our branch to 2021.04.. shouldn't be much longer.

**Reply by ltrsunny** [reply]
> > Thank you very much. I also want to contribute something if I can.
> 
> Once I have our u-boot updated to the current master you can try writing the GPIO driver. I don't think it'll be too difficult and I can help out. There isn't much point starting that now as a lot of things have changed since 2021.01 (our version) so you'd end up having to do fixes to get it to work again after we are running master.
> 
> I'm close to getting our branch to 2021.04.. shouldn't be much longer.

Sure！

**Reply by fifteenhex** [reply]
I've gotten a master based branch running now. https://github.com/linux-chenxing/u-boot/tree/mstar_rebase_mainline_20211217
I haven't tested ethernet yet.

**Reply by ltrsunny** [reply]
Thanks! Let me test tomorrow!

**Reply by ltrsunny** [reply]
> I've gotten a master based branch running now. https://github.com/linux-chenxing/u-boot/tree/mstar_rebase_mainline_20211217 I haven't tested ethernet yet.

Ethernet works fine, but `mmcinfo` (or `mmc part`) still gets 
```
timeout while polling
timeout while polling
timeout while polling
timeout while polling
timeout while polling
```

**Reply by fifteenhex** [reply]
I'm working on it...

**Reply by fifteenhex** [reply]
```
=> mmc info 
hardware reset
timeout while polling, status: 0x0, intflags 0x2
Device: MSC
Manufacturer ID: 27
OEM: 5048
Name: SD32G 
Bus Speed: 50000000
Mode: SD High Speed (50MHz)
Rd Block Len: 512
SD version 3.0
High Capacity: Yes
Capacity: 28.9 GiB
Bus Width: 4-bit
Erase Group Size: 512 Bytes
=> 
```

Got it sort of working. I'll push some changes soon..

**Reply by ltrsunny** [reply]
Thanks

**Reply by fifteenhex** [reply]
SD seems to be working now (I tested on a different board but they are the same). If you force pull from the u-boot branch you will get those changes.
I'm going to work on the USB over the next few days as it'll help me with boards that don't have ethernet.

**Reply by ltrsunny** [reply]
Thanks for your great job!

**Reply by fifteenhex** [reply]
Basic GPIO driver has been added too. pin numbering is the same as linux so use gpioinfo there to get the pin number you need.

**Reply by ltrsunny** [reply]
Thank you very much

**Reply by fifteenhex** [reply]
I have made some progress on USB in u-boot too. It isn't seeing any device yet but u-boot is operating the USB controller.

**Reply by ltrsunny** [reply]
> I have made some progress on USB in u-boot too. It isn't seeing any device yet but u-boot is operating the USB controller.

Does USB gadget mode work over the same driver?

**Reply by fifteenhex** [reply]
I don't think. The USB gadget in the SigmaStar is an MUSB controller I think. I'm not sure if the SSD202D has the gadget controller.

**Reply by ltrsunny** [reply]
> I don't think. The USB gadget in the SigmaStar is an MUSB controller I think. I'm not sure if the SSD202D has the gadget controller.

Emm… let me try to ask the vendor.

**Reply by ltrsunny** [reply]
The module's vendor said it is not integrated in the chip

**Comment by fifteenhex** [comment]
Note:

For usb we need to devicetree'ify: https://github.com/linux-chenxing/u-boot/blob/mstar_rebase_mainline_20211217/drivers/usb/host/ehci-faraday.c
And add the USB PLL driver..
